worker/worker.go

```diff
@@ -792,16 +842,69 @@ func joinErrors(errs []error) error {
 	}
 }
 
+func (w *worker) hostPriceTableByContract(ctx context.Context, hk consensus.PublicKey, hostIP string, revision *types.FileContractRevision) (pt rhp.HostPriceTable, err error) {
+	// TODO: Check if valid price table exists and handle the following edge cases.
+	// - If no price table exists, we fetch a new one.
+	// - If a price table exists that is valid but about to expire use it but launch an update.
+	// - If a price table exists that is valid and doesn't expire soon, use it.
+
+	// Get price table.
+	err = w.withTransportV3(ctx, hostIP, hk, func(t *rhpv3.Transport) (err error) {
+		// The FundAccount RPC requires a SettingsID, which we also have to pay
+		// for. To simplify things, we pay for the SettingsID using the full
+		// amount, with the "refund" going to the desired account; we then top
+		// up the account to cover the cost of the two RPCs.
+		pt, err = rhpv3.RPCPriceTable(t, func(priceTable rhpv3.HostPriceTable) (rhpv3.PaymentMethod, error) {
+			cost := priceTable.UpdatePriceTableCost.Add(priceTable.FundAccountCost)
+
+			// TODO: gouging check on price table
+
+			refundAccount := rhp.Account(w.deriveAccountKey(hk).PublicKey())
+			rk := w.deriveRenterKey(hk)
+			payment, ok := rhpv3.PayByContract(revision, cost, refundAccount, rk)
+			if !ok {
+				return nil, errors.New("insufficient funds")
+			}
+			return &payment, nil
+		})
+		return err
+	})
+
+	// TODO: update cached price table.
+	return pt, err
+}
+
+func (w *worker) preparePayment(hk consensus.PublicKey, amt types.Currency) rhpv3.PayByEphemeralAccountRequest {
+	pk := w.deriveAccountKey(hk)
+	return rhpv3.PayByEphemeralAccount(rhpv3.Account(pk.PublicKey()), amt, math.MaxUint64, pk)
+}
+
+func (w *worker) accountsHandlerGET(jc jape.Context) {
+	accounts, err := w.accounts.All()
+	if jc.Check("failed to fetch accounts", err) != nil {
+		return
+	}
+	jc.Encode(accounts)
+}
+
 // New returns an HTTP handler that serves the worker API.
-func New(masterKey [32]byte, b Bus, sessionTTL time.Duration) http.Handler {
+func New(masterKey [32]byte, b Bus, sessionTTL time.Duration) (http.Handler, error) {
 	w := &worker{
+		id:        "worker", // TODO: make this configurable for multi-worker setup.
 		bus:       b,
 		pool:      newSessionPool(sessionTTL),
 		masterKey: masterKey,
+		accounts:  nil,
+	}
+	w.accounts = &accounts{
+		accounts: make(map[rhpv3.Account]*account),
+		w:        w,
```

Comment:
circular references are kinda sus -- can we give `accounts` only the information it needs (`masterKey`?), rather than the entire worker?